pkg/provider/kind/kind.go

```diff
@@ -0,0 +1,215 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kind
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	k8sProvider "github.com/prometheus/test-infra/pkg/provider/k8s"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"fmt"
+	"github.com/prometheus/test-infra/pkg/provider"
+	"gopkg.in/alecthomas/kingpin.v2"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/util/homedir"
+	"log"
+	"sigs.k8s.io/kind/pkg/cluster"
+	"sigs.k8s.io/kind/pkg/cmd"
+	"strings"
+)
+
+// New is the KIND constructor.
+
+type Resource = provider.Resource
+
+// GKE holds the fields used to generate an API request.
+type KIND struct {
+
+	// The k8s provider used when we work with the manifest files.
+	k8sProvider *k8sProvider.K8s
+	// The kind provider used to instantiate a new provider.
+	kindProvider *cluster.Provider
+	// Final DeploymentFiles files.
+	DeploymentFiles []string
+	// Final DeploymentVars.
+	DeploymentVars map[string]string
+	// DeployResource to construct DeploymentVars and DeploymentFiles
+	DeploymentResource *provider.DeploymentResource
+	// Content bytes after parsing the template variables, grouped by filename.
+	kindResources []Resource
+	// K8s resource.runtime objects after parsing the template variables, grouped by filename.
+	k8sResources []k8sProvider.Resource
+
+	ctx context.Context
+}
+
+func New(dr *provider.DeploymentResource) *KIND {
+	return &KIND{
+		DeploymentResource: dr,
+		kindProvider: cluster.NewProvider(
+			cluster.ProviderWithLogger(cmd.NewLogger()),
+		),
+		ctx: context.Background(),
+	}
+}
+
+// SetupDeploymentResources Sets up DeploymentVars and DeploymentFiles
+func (c *KIND) SetupDeploymentResources(*kingpin.ParseContext) error {
+	customDeploymentVars := map[string]string{
+		"NGINX_SERVICE_TYPE":        "NodePort",
+		"LOADGEN_SCALE_UP_REPLICAS": "2",
+	}
+
+	c.DeploymentFiles = c.DeploymentResource.DeploymentFiles
+	c.DeploymentVars = provider.MergeDeploymentVars(
+		c.DeploymentResource.DefaultDeploymentVars,
+		c.DeploymentResource.FlagDeploymentVars,
+		customDeploymentVars,
+	)
+	return nil
+}
+
+// KINDDeploymentsParse parses the environment/kind deployment files and saves the result as bytes grouped by the filename.
+// Any variables passed to the cli will be replaced in the resources files following the golang text template format.
+func (c *KIND) KINDDeploymentsParse(*kingpin.ParseContext) error {
+	deploymentResource, err := provider.DeploymentsParse(c.DeploymentFiles, c.DeploymentVars)
+	if err != nil {
+		log.Fatalf("Couldn't parse deployment files: %v", err)
```

Comment:
We were trying to get rid of `log.Fatalf` wherever possible. it'll be nice if we would just return the error.  (for gke and some other tools they are still there because nobody got around removing them)

https://dave.cheney.net/2015/11/05/lets-talk-about-logging